
#' Extract contiguous co-methylated genomic regions from a list of
#' pre-defined genomic regions
#'
#' @param betaMatrix matrix of beta values, with row names = CpG IDs,
#'    column names = sample IDs. This is typically genome-wide methylation beta values.
#'
#' @param file name of input file with clusters of CpG locations (i.e. CpGs located closely
#'  to each other on the genome). This file can be generated by the
#'  \code{WriteCloseByAllRegions} function.
#'
#' @param fileType file extension for input file, can be "gmt" or "RDS"
#'
#' @param arrayType Type of array, can be "450k" or "EPIC"
#'
#' @param returnAllCpGs When there is not a contiguous comethylated region in the inputing
#' pre-defined region, \code{returnAllCpGs = 1} indicates outputting all the CpGs in the input
#' regions, \code{returnAllCpGs = 0} indicates not returning any CpG.
#'
#' @param ...
#'
#' @return A list of two components:
#'
#' \itemize{
#'   \item{\code{Contiguous_Regions} : }{A data frame with \code{CpG} (CpG name), \code{Chr} (chromosome
#'    number), \code{MAPINFO} (genomic position), \code{r_drop}(correlation between the CpG with rest
#'   of the CpGs), \code{keep} (indicator for co-methylated CpG), \code{keep_contiguous} (index for
#'   contiguous comethylated subregions)
#'   }
#'
#'   \item{\code{CpGs_subregions} : } {lists of CpGs in each contiguous co-methylated
#'    subregion from all the regions
#'    }
#'
#' @export
#'
#' @examples
#'
#' \dontrun{
#'    data(betaMatrixChr22_df)
#'
#'    CoMethAllRegions(
#'       betaMatrix = betaMatrixChr22_df,
#'
#'       file = system.file(
#'          "extdata",
#'          "CpGislandsChr22_ex.RDS",
#'          package = 'coMethDMR',
#'          mustWork = TRUE),
#'
#'       fileType = "RDS",
#'       arrayType = "450k",
#'       returnAllCpGs = FALSE
#'
#'       )
#'}
CoMethAllRegions <- function(betaMatrix,
                             file, fileType = c("gmt","RDS"),
                             arrayType = c("450k","EPIC"),
                             returnAllCpGs = FALSE,
                             ...){

  arrayType <- match.arg(arrayType)
  fileType  <- match.arg(fileType)

  ### Read file of close by CpGs ###
  switch(
    fileType,
    "RDS" = {
      closeByGenomicRegion_ls <- readRDS(file)
    },
    "gmt" = {
      closeByGenomicRegion_ls <- read_gmt(file)
    }
  )


  ### Extract contiguous comethylated region(s) from each close by region ###
  # A pathway is a set of CpGs in a region
  coMethCpGsAllREgions_ls <- lapply(
    unname(closeByGenomicRegion_ls$pathways),
    FUN = CoMethSingleRegion,
    betaMatrix, arrayType, returnAllCpGs
  )


  ### Return list of cotiguous comethylated CpGs by Regions ###
  out_ContigRegions <- (lapply(coMethCpGsAllREgions_ls, `[[`, 1))
  out_coMethCpGsAll <- unlist(
    lapply(coMethCpGsAllREgions_ls, `[[`, 2),
    recursive = FALSE
  )


  ### return output ###
  list(
    CpGsSubregions    = out_coMethCpGsAll,
    contiguousRegions = out_ContigRegions
  )

}
